### ISSUE TITLE
Attempt to render classical modular form page gets stuck in an infinite loop (sage bug?)

### ISSUE THREAD

**Issue by AndrewVSutherland** [LMFDB/lmfdb#1505]
Go to the elliptic curve isogeny class http://www.lmfdb.org/EllipticCurve/Q/234446/a/ and click on the modular form link http://www.lmfdb.org/ModularForm/GL2/Q/holomorphic/234446/2/1/a/.

This should result in a message saying the form is not in the database, but instead the page hangs and eventually times out with a 502 server error.  Similar behavior occurs for several other (but by no means all) elliptic curve isogeny classes.

If you run the lmfdb server locally, it appears to go into an infinite loop (or at least a multi-hour computation) inside sage while attempting to process this request (it is unclear to me why it should be accessing Sage at all, it only needs to determine that the form is not in the database).

Pressing ctrl-C yields an apparently infinite traceback, a portion of which appears below:

```
Traceback (most recent call last):
  File "/usr/local/sage/sage-1/local/lib/python/SocketServer.py", line 295, in _handle_request_noblock
    self.process_request(request, client_address)
  File "/usr/local/sage/sage-1/local/lib/python/SocketServer.py", line 321, in process_request
    self.finish_request(request, client_address)
  File "/usr/local/sage/sage-1/local/lib/python/SocketServer.py", line 334, in finish_request
    self.RequestHandlerClass(request, client_address, self)
  File "/usr/local/sage/sage-1/local/lib/python/SocketServer.py", line 655, in __init__
    self.handle()
  File "/usr/local/sage/sage-1/local/lib/python2.7/site-packages/Werkzeug-0.11.4-py2.7.egg/werkzeug/serving.py", line 216, in handle
    rv = BaseHTTPRequestHandler.handle(self)
  File "/usr/local/sage/sage-1/local/lib/python/BaseHTTPServer.py", line 340, in handle
    self.handle_one_request()
  File "/usr/local/sage/sage-1/local/lib/python2.7/site-packages/Werkzeug-0.11.4-py2.7.egg/werkzeug/serving.py", line 251, in handle_one_request
    return self.run_wsgi()
  File "/usr/local/sage/sage-1/local/lib/python2.7/site-packages/Werkzeug-0.11.4-py2.7.egg/werkzeug/serving.py", line 193, in run_wsgi
    execute(self.server.app)
  File "/usr/local/sage/sage-1/local/lib/python2.7/site-packages/Werkzeug-0.11.4-py2.7.egg/werkzeug/serving.py", line 181, in execute
    application_iter = app(environ, start_response)
  File "/usr/local/sage/sage-1/local/lib/python2.7/site-packages/Flask-0.10.1-py2.7.egg/flask/app.py", line 1836, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/local/sage/sage-1/local/lib/python2.7/site-packages/Flask-0.10.1-py2.7.egg/flask/app.py", line 1817, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/local/sage/sage-1/local/lib/python2.7/site-packages/Flask-0.10.1-py2.7.egg/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/sage/sage-1/local/lib/python2.7/site-packages/Flask-0.10.1-py2.7.egg/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/drew/github/lmfdb/lmfdb/modular_forms/elliptic_modular_forms/views/emf_main.py", line 130, in render_elliptic_modular_forms
    return render_web_newform(**info)
  File "/home/drew/github/lmfdb/lmfdb/modular_forms/elliptic_modular_forms/views/emf_render_web_newform.py", line 42, in render_web_newform
    info = set_info_for_web_newform(level, weight, character, label, **kwds)
  File "/home/drew/github/lmfdb/lmfdb/modular_forms/elliptic_modular_forms/views/emf_render_web_newform.py", line 71, in set_info_for_web_newform
    WNF = WebNewForm_cached(level=level, weight=weight, character=character, label=label)
  File "/home/drew/github/lmfdb/lmfdb/modular_forms/elliptic_modular_forms/backend/web_newforms.py", line 965, in WebNewForm_cached
    F = WebNewForm(level,weight,character,label,**kwds)
  File "sage/misc/classcall_metaclass.pyx", line 329, in sage.misc.classcall_metaclass.ClasscallMetaclass.__call__ (build/cythonized/sage/misc/classcall_metaclass.c:1188)
    return cls.classcall(cls, *args, **kwds)
  File "sage/misc/cachefunc.pyx", line 973, in sage.misc.cachefunc.CachedFunction.__call__ (build/cythonized/sage/misc/cachefunc.c:5376)
    w = self.f(*args, **kwds)
  File "/usr/local/sage/sage-1/local/lib/python2.7/site-packages/sage/structure/unique_representation.py", line 1021, in __classcall__
    instance = typecall(cls, *args, **options)
  File "sage/misc/classcall_metaclass.pyx", line 496, in sage.misc.classcall_metaclass.typecall (build/cythonized/sage/misc/classcall_metaclass.c:1614)
    return (<PyTypeObject*>type).tp_call(cls, args, kwds)
  File "/home/drew/github/lmfdb/lmfdb/modular_forms/elliptic_modular_forms/backend/web_newforms.py", line 516, in __init__
    else False),
  File "/home/drew/github/lmfdb/lmfdb/modular_forms/elliptic_modular_forms/backend/web_character.py", line 233, in __init__
    self._default_value = WebChar_cached(modulus, number, update_from_db=True, compute=True)
  File "/home/drew/github/lmfdb/lmfdb/modular_forms/elliptic_modular_forms/backend/web_character.py", line 273, in WebChar_cached
    X = WebChar(modulus,number,**kwds)
  File "sage/misc/classcall_metaclass.pyx", line 329, in sage.misc.classcall_metaclass.ClasscallMetaclass.__call__ (build/cythonized/sage/misc/classcall_metaclass.c:1188)
    return cls.classcall(cls, *args, **kwds)
  File "sage/misc/cachefunc.pyx", line 973, in sage.misc.cachefunc.CachedFunction.__call__ (build/cythonized/sage/misc/cachefunc.c:5376)
    w = self.f(*args, **kwds)
  File "/usr/local/sage/sage-1/local/lib/python2.7/site-packages/sage/structure/unique_representation.py", line 1021, in __classcall__
    instance = typecall(cls, *args, **options)
  File "sage/misc/classcall_metaclass.pyx", line 496, in sage.misc.classcall_metaclass.typecall (build/cythonized/sage/misc/classcall_metaclass.c:1614)
    return (<PyTypeObject*>type).tp_call(cls, args, kwds)
  File "/home/drew/github/lmfdb/lmfdb/modular_forms/elliptic_modular_forms/backend/web_character.py", line 92, in __init__
    update_from_db=update_from_db
  File "/home/drew/github/lmfdb/lmfdb/modular_forms/elliptic_modular_forms/backend/web_object.py", line 344, in __init__
    self.init_dynamic_properties()
  File "/home/drew/github/lmfdb/lmfdb/modular_forms/elliptic_modular_forms/backend/web_character.py", line 134, in init_dynamic_properties
    self.sage_character = self.character.sage_character()
  File "dirichlet_conrey.pyx", line 1475, in dirichlet_conrey.DirichletCharacter_conrey.sage_character (dirichlet_conrey.c:19264)
  File "dirichlet_conrey.pyx", line 395, in dirichlet_conrey.DirichletGroup_conrey.standard_dirichlet_group (dirichlet_conrey.c:6601)
  File "sage/structure/factory.pyx", line 361, in sage.structure.factory.UniqueFactory.__call__ (build/cythonized/sage/structure/factory.c:1469)
    key, kwds = self.create_key_and_extra_args(*args, **kwds)
  File "sage/structure/factory.pyx", line 459, in sage.structure.factory.UniqueFactory.create_key_and_extra_args (build/cythonized/sage/structure/factory.c:2602)
    return self.create_key(*args, **kwds), {}
  File "/usr/local/sage/sage-1/local/lib/python2.7/site-packages/sage/modular/dirichlet.py", line 1903, in create_key
    base_ring = rings.CyclotomicField(e)
  File "sage/structure/factory.pyx", line 363, in sage.structure.factory.UniqueFactory.__call__ (build/cythonized/sage/structure/factory.c:1551)
    return self.get_object(version, key, kwds)
  File "sage/structure/factory.pyx", line 406, in sage.structure.factory.UniqueFactory.get_object (build/cythonized/sage/structure/factory.c:1906)
    obj = self.create_object(version, key, **extra_args)
  File "/usr/local/sage/sage-1/local/lib/python2.7/site-packages/sage/rings/number_field/number_field.py", line 1098, in create_object
    return NumberField_cyclotomic(n, names, embedding=embedding)
  File "/usr/local/sage/sage-1/local/lib/python2.7/site-packages/sage/rings/number_field/number_field.py", line 8895, in __init__
    maximize_at_primes=maximize_at_primes)
  File "/usr/local/sage/sage-1/local/lib/python2.7/site-packages/sage/rings/number_field/number_field.py", line 6444, in __init__
    assume_disc_small=assume_disc_small, maximize_at_primes=maximize_at_primes, structure=structure)
  File "/usr/local/sage/sage-1/local/lib/python2.7/site-packages/sage/rings/number_field/number_field.py", line 1223, in __init__
    embedding = number_field_morphisms.create_embedding_from_approx(self, embedding)
  File "sage/rings/number_field/number_field_morphisms.pyx", line 494, in sage.rings.number_field.number_field_morphisms.create_embedding_from_approx (build/cythonized/sage/rings/number_field/number_field_morphisms.c:7033)
    if not P.is_exact() or f(gen_image) != 0:
  File "sage/structure/element.pyx", line 1007, in sage.structure.element.Element.__richcmp__ (build/cythonized/sage/structure/element.c:9714)
    return coercion_model.richcmp(self, other, op)
  File "sage/structure/coerce.pyx", line 1801, in sage.structure.coerce.CoercionModel_cache_maps.richcmp (build/cythonized/sage/structure/coerce.c:18652)
    return PyObject_RichCompare(x, y, op)
  File "sage/structure/element.pyx", line 1005, in sage.structure.element.Element.__richcmp__ (build/cythonized/sage/structure/element.c:9690)
    return (<Element>self)._richcmp_(<Element>other, op)
  File "sage/structure/element.pyx", line 1040, in sage.structure.element.Element._richcmp_ (build/cythonized/sage/structure/element.c:9871)
    c = left._cmp_(right)
  File "sage/rings/real_lazy.pyx", line 699, in sage.rings.real_lazy.LazyFieldElement._cmp_ (build/cythonized/sage/rings/real_lazy.c:8630)
    left, right = self.approx(), other.approx()
  File "sage/rings/real_lazy.pyx", line 771, in sage.rings.real_lazy.LazyFieldElement.approx (build/cythonized/sage/rings/real_lazy.c:9311)
    return self.eval(self._parent.interval_field())
  File "sage/rings/real_lazy.pyx", line 1133, in sage.rings.real_lazy.LazyBinop.eval (build/cythonized/sage/rings/real_lazy.c:12492)
    left = self._left.eval(R)
  File "sage/rings/real_lazy.pyx", line 1133, in sage.rings.real_lazy.LazyBinop.eval (build/cythonized/sage/rings/real_lazy.c:12492)
    left = self._left.eval(R)
  File "sage/rings/real_lazy.pyx", line 1133, in sage.rings.real_lazy.LazyBinop.eval (build/cythonized/sage/rings/real_lazy.c:12492)
...
```

**Comment by sehlen** [comment]
I'll look into it.

**Comment by sehlen** [comment]
I probably even know what’s going on which is 90% solved in a branch I haven’t made a pull request for and something that I didn’t even think would affect the website currently. The WebChar class computes way too much unused (currently, on the website) stuff on the fly - something I’ve seen when looking into your question if the corresponding collection is used. my changes already make the page load now but still we have to wait way too long (maybe 20s) and the reason is still the character (the trivial one -haha), so I will fix it and submit a pull request tonight.

**Comment by JohnCremona** [comment]
I fixed this (but see below) in Birstol by asking @fredstro to provide a function to check if the modular form is in the database, and when it is not, not offering the modular form in the Related Objects box for either the curve or the isogeny class.  **But** more recently I added the modular form link in the title of the Modular Form section on the main page, without doing the check.  So it is all my fault, and I will fix it: I suggest by always having the label of the modular form in the title of that section, but only making it a link when the form is in the database.

**Comment by JohnCremona** [comment]
The top-level issue has been fixed in #1508 but others may want to work out why the failure was so messy instead of calmly showing the "This form is not in the database" page.  I suggest that this issue not be closed when #1508 is settled, only when this underlying problem is.

**Comment by sehlen** [comment]
@JohnCremona Thanks for fixing it on the elliptic curve side! I already made the page load with #1507, which you might have not noticed since I didn't comment on the issue but only referenced it in the pull request (which is shown on this page but not nearly as clear as the comments).
